### PR TITLE
chore: Optimize WAL segment writes to object storage

### PR DESCRIPTION
**Pull request by benclive**
**What this PR does / why we need it**:
Two optimisations:
* Use multiple flush threads for object storage
* Use buffered writer for putting data in the pipe.
* The buffered writer is needed to avoid sending data to object storage with small packets, as writing from the WAL segment happens in small bursts.

### CODE REVIEW

**Review by cyriltovena — APPROVED** [review]
LGTM

File: pkg/ingester-rf1/objstore/storage.go [hunk 1]
```diff
@@ -31,6 +31,10 @@ func New(
 	storageConfig storage.Config,
 	clientMetrics storage.ClientMetrics,
 ) (*Multi, error) {
+	if storageConfig.GCSConfig.ChunkBufferSize == 0 {
```
Comment (benclive) [inline pkg/ingester-rf1/objstore/storage.go:34]: note: we're already passing by value here so this config is not the same struct as in the top level modules.go

Comment (cyriltovena) [reply pkg/ingester-rf1/objstore/storage.go:34]: storageConfig yes but are we sure this is the case for `GCSConfig` ?

File: pkg/ingester-rf1/ingester.go [hunk 2]
```diff
@@ -590,7 +592,8 @@ func (i *Ingester) doFlushTick() {
 	// TODO: use multiple flush queues if required
 	// Don't write empty segments if there is nothing to write.
 	if currentFlushCtx.segmentWriter.InputSize() > 0 {
-		i.flushQueues[0].Enqueue(currentFlushCtx)
+		i.flushQueues[i.nextFlushQueue].Enqueue(currentFlushCtx)
+		i.nextFlushQueue = (i.nextFlushQueue + 1) % len(i.flushQueues)
```
Comment (cstyan) [inline pkg/ingester-rf1/ingester.go:596]: how do you know it's safe to loop around to the beginning of the `flushQueues` again? do we not need to locate data for the same steam/segment next to eachother?

**Pull request closed without merging**